### PR TITLE
Remove ThingScreenlet

### PR DESCRIPTION
**Pull request by phcp**
We need to move the thingScreenlet to the sample to keep it working without using screens. In a second moment, I guess we need to rethink about a small sample just to exemplify the consumer possibilities.

### CODE REVIEW

**Comment by phcp** [comment]
cc @nhpatt @victorg1991

File: library/src/main/kotlin/com/liferay/apio/consumer/model/Thing.kt [hunk 1]
```diff
@@ -16,11 +16,22 @@ package com.liferay.apio.consumer.model
 
 import com.liferay.apio.consumer.ApioException
 import com.liferay.apio.consumer.graph
+import com.liferay.apio.consumer.requestProperties
 
-data class Thing(val id: String, val type: List<String>, val attributes: Map<String, Any>, val name: String? = null)
+
+typealias Type = List<String>
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/apio/consumer/model/Thing.kt]: Can you rename to something clearer? I know it's the Type of the Thing but Type is a work with so many meanings...

File: library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 2]
```diff
@@ -47,6 +38,86 @@ fun fetch(
...
+	} ?: onComplete(Result.of(null, { ApioException("Thing not found") }))
+}
+
+fun performOperationRequest(url: String, method: String, attributes: Map<String, Any>,
+	onComplete: (Result<Response, Exception>) -> Unit) {
+	launch(UI) {
+		async(CommonPool) {
+
+			val json = Gson().toJson(attributes)
+			val request = createRequest(HttpUrl.parse(url), credential).newBuilder()
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: maybe we should extract the request code to a common class/function, I'm not sure.

File: library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 3]
```diff
@@ -47,6 +38,86 @@ fun fetch(
...
+		}.await().let(onComplete)
+	}
+}
+
+fun requestProperties(url: String, onComplete: (List<Property>) -> Unit) {
+	val request = createRequest(HttpUrl.parse(url), credential)
+
+	launch(UI) {
+		async(CommonPool) {
+			val okHttpClient = OkHttpClient()
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: And the code that interacts with OkHttpClient

File: library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 4]
```diff
@@ -71,16 +142,9 @@ private fun request(url: HttpUrl,
 	val request = createRequest(httpUrl, credential)
 
 	val okHttpClient = OkHttpClient()
-	if (BuildConfig.DEBUG) {
-		IdlingResources.registerOkHttp(okHttpClient, httpUrl.toString())
-	}
 
 	val execute = okHttpClient.newCall(request).execute()
 
-	if (BuildConfig.DEBUG) {
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: ? :(

File: build.gradle [hunk 5]
```diff
@@ -86,7 +86,7 @@ ext.libraries = [appCompat          : 'com.android.support:appcompat-v7:' + supp
                  okhttp             : 'com.squareup.okhttp3:okhttp:3.10.0',
                  result             : 'com.github.kittinunf.result:result:1.3.0',
                  anko_commons       : 'org.jetbrains.anko:anko-commons:0.10.4',
-                 picasso            : 'com.squareup.picasso:picasso:2.71828',
+                 picasso            : 'com.squareup.picasso:picasso:2.5.2',
```
Comment (nhpatt) [inline build.gradle]: Why? a bug?

File: apio-consumer/src/debug/kotlin/com/liferay/apio/consumer/IdlingResources.kt [hunk 6]
```diff
@@ -12,8 +12,8 @@ object IdlingResources {
 	}
 
 	fun unregisterOkHttp(okHttpClient: OkHttpClient, name: String) {
-//		val create = OkHttp3IdlingResource.create("_", okHttpClient)
-//		IdlingRegistry.getInstance().unregister(create)
+		val create = OkHttp3IdlingResource.create("_", okHttpClient)
```
Comment (nhpatt) [inline apio-consumer/src/debug/kotlin/com/liferay/apio/consumer/IdlingResources.kt]: Is this method called now?

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+	}
+}
+
+private fun request(url: HttpUrl,
+	fields: Map<String, List<String>>,
+	embedded: List<String>, credentials: String?): Response {
+	val httpUrl = createUrl(url, fields, embedded)
+
+	if (credentials != null) {
+		credential = credentials
```
Comment (nhpatt) [inline apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: the mixture of credential and credentials is strange

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+		val form = expects?.let { OperationForm(it) }
+
+		id to Operation(id, target, type, method, form)
+	}.toMap().toMutableMap()
+}
+
+private fun parseType(type: Any?): List<String> {
+	return (type as? String)?.let { listOf(it) } ?:
+	type as? List<String> ?: listOf()
```
Comment (nhpatt) [inline apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: Can you use the typealias?

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+typealias FoldedAttributes = Pair<MutableMap<String, Any>, MutableMap<String, Thing?>>
+
+private fun foldEntry(context: Context?) = { acc: FoldedAttributes, entry: Entry<String, Any> ->
+	val (attributes, things) = acc
+
+	val key = entry.key
+	val value = entry.value
+
+	when {
+		value is Map<*, *> -> (value as? Map<String, Any>)?.apply {
```
Comment (nhpatt) [inline apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: we should extract functions here to know what we are checking

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+				}
+			}
+			else {
+				val (attr, embeddedThings) = foldTree(this, context)
+
+				things.putAll(embeddedThings)
+				attributes[key] = attr
+			}
+		}
+		value is List<*> && !value.filterIsInstance<Map<String, Any>>().isEmpty() ->
```
Comment (nhpatt) [inline apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: extract functions and flatten to be at the same level of the switch

File: library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt (hunk 4 above)
Comment (victorg1991) [reply library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: No Android, no views, so no need for idling resources, right? 🤔

File: build.gradle (hunk 5 above)
Comment (phcp) [reply build.gradle]: There was a breaking changing, but I fixed it and kept the updated version.

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt (hunk 8 above)
Comment (phcp) [reply apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: Done!

**Comment by phcp** [comment]
@nhpatt I only moved the changes from Android-SDK, maybe this refactoring could be done in another PR as @victorg1991 said. I totally agree that there are codes to improve and, of course, we can contribute with that since we will be working on Swift consumer also.

**Comment by nhpatt** [comment]
Yep, It's ok for me :)

**Comment by phcp** [comment]
Great! I will do some small changes in this PR as you suggested and then update here.

**Comment by phcp** [comment]
@victorg1991 About the Thing Screenlet I think that it's better to do a PR with this before the Forms Screenlet PR, I will prepare it here.

File: library/src/main/kotlin/com/liferay/apio/consumer/model/Thing.kt (hunk 1 above)
Comment (phcp) [reply library/src/main/kotlin/com/liferay/apio/consumer/model/Thing.kt]: Yep, I agree, changed to ThingType.

File: apio-consumer/src/debug/kotlin/com/liferay/apio/consumer/IdlingResources.kt (hunk 6 above)
Comment (phcp) [reply apio-consumer/src/debug/kotlin/com/liferay/apio/consumer/IdlingResources.kt]: I don't think so, they are unused in the project. So I just removed them, it's better than keep it commented. Ok?

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt (hunk 7 above)
Comment (phcp) [reply apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: Yes, it's strange, I renamed the property to `credentials` and also renamed the parameter to `newCredentials`.

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt (hunk 8 above)
Comment (phcp) [reply apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: Yes, it's using ThingType now.

File: apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt (hunk 9 above)
Comment (phcp) [reply apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt]: I extracted in two functions, the PR is now updated.

**Comment by allanragec** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by allanragec** [comment]
Thank you, pull request merged! [See changes here.](https://github.com/liferay-mobile/apio-consumer-android/compare/8c73c9a~2...8c73c9a) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Pull request merged**